collect/collect.go

```diff
@@ -1354,57 +1355,60 @@ func (i *InMemCollector) signalDroppedTraceDecisions(ctx context.Context, msg st
 		i.Logger.Warn().Logf("dropped trace decision channel is full. Dropping message")
 	}
 }
+func (i *InMemCollector) processTraceDecisions(msg string, decisionType decisionType) {
 
-func (i *InMemCollector) processDropDecisions(msg string) {
-	i.Metrics.Increment("drop_decisions_received")
+	i.Metrics.Increment(fmt.Sprintf("%s_decisions_received", decisionType.String()))
+	if len(msg) == 0 {
+		return
+	}
 
-	ids := newDroppedTraceDecision(msg)
+	// Deserialize the message into trace decisions
+	decisions := make([]TraceDecision, 0)
+	var err error
+	switch decisionType {
+	case keptDecision:
+		decisions, err = newKeptTraceDecision(msg)
+		if err != nil {
+			i.Logger.Error().Logf("Failed to unmarshal kept trace decision message. %s", err)
+			return
+		}
+	case dropDecision:
+		decisions, err = newDroppedTraceDecision(msg)
+		if err != nil {
+			i.Logger.Error().Logf("Failed to unmarshal drop trace decision message. %s", err)
+			return
+		}
+	default:
+		i.Logger.Error().Logf("unknown decision type %s while processing trace decisions", decisionType)
+		return
+	}
 
-	if len(ids) == 0 {
+	if len(decisions) == 0 {
 		return
 	}
 
 	toDelete := generics.NewSet[string]()
-	for _, id := range ids {
-
-		trace := i.cache.Get(id)
-		// if we don't have the trace in the cache, we don't need to do anything
+	for _, decision := range decisions {
+		// Assume TraceDecision implements a common interface like TraceID
+		trace := i.cache.Get(decision.TraceID)
 		if trace == nil {
-			i.Logger.Debug().Logf("trace not found in cache for trace decision")
+			i.Logger.Debug().Logf("trace not found in cache for %s decision", decisionType.String())
 			continue
 		}
-		toDelete.Add(id)
-
-		i.sampleTraceCache.Record(trace, false, "")
-
-	}
-
-	i.cache.RemoveTraces(toDelete)
-}
+		toDelete.Add(decision.TraceID)
 
-func (i *InMemCollector) processKeptDecision(msg string) {
-	i.Metrics.Increment("kept_decisions_received")
+		if decisionType == keptDecision {
+			trace.SetSampleRate(decision.SampleRate)
+			trace.KeepSample = decision.Kept
+		}
 
-	td, err := newKeptTraceDecision(msg)
-	if err != nil {
-		i.Logger.Error().Logf("Failed to unmarshal trace decision message. %s", err)
-		return
-	}
+		i.sampleTraceCache.Record(trace, decision.Kept, decision.KeptReason)
 
-	toDelete := generics.NewSet[string]()
-	trace := i.cache.Get(td.TraceID)
-	// if we don't have the trace in the cache, we don't need to do anything
-	if trace == nil {
-		i.Logger.Debug().Logf("trace not found in cache for trace decision")
-		return
+		// Send the processed trace (only in Kept case)
```

Comment:
I realized that we should still call `send` for dropped traces since the `trace_send_dropped` metric is in there and it does check against `trace.Kept` inside `send` anyway